docs/flip/1777-default-dtype.md

```diff
@@ -0,0 +1,102 @@
+
+The current behavior is problematic and results in silent bugs especially for dtypes that do not fit inside float32 (complex, float64). Also the linen dtype behavior is significantly different from how NumPy and by extension JAX handle dtypes.
+
+
+### Dtypes in JAX
+
+JAX uses a NumPy inspired [dtype promotion](https://github.com/google/jax/blob/main/jax/_src/dtypes.py) mechanism as explained [here](https://jax.readthedocs.io/en/latest/type_promotion.html?highlight=lattice#type-promotion-semantics). The type promotion rules are summarized by the following type lattice:
+
+![JAX type promotion lattice](https://jax.readthedocs.io/en/latest/_images/type_lattice.svg)
+
+
+
+## Dtypes in Linen
+
+Beside input arguments, state and in particular parameters could affect dtype promotion. For example: We might feed a float64 input to a Dense layer with float32 parameters. Currently the result would be truncated to float32. If the input is a complex number the result is even worse because the imaginary part will be silently dropped when casting to float32.
+
+By using the dtype promotion rules already available in JAX we can avoid this issue. A public api is available called `jax.numpy.result_dtype(*args)` which returns the dtype that JAX would promote the given arguments in accordance with the type promotion lattice. For linen layers the arguments would be the layer inputs together with the parameters so for example this would be inputs, kernel, and bias for a linear layer.
+
+
+# Implementation
+
+A simplified example implementation:
+
+
+```python
+def promote_arrays(*xs, dtype):
+ if dtype is None:
+   dtype = jnp.result_type(*jax.tree_leaves(xs))
+ return jax.tree_map(lambda x: jnp.asarray(x, dtype), xs)
+
+Dtype = Any
+class Dense(nn.Module):
+ features: int
+ kernel_init: Callable
+ bias_init: Callable
+ dtype: Optional[Dtype] = None
+ param_dtype: Dtype = jnp.float32
+
+ @nn.compact
+ def __call__(self, x):
+   kernel = self.param("kernel", 
+                       self.kernel_init,
+                       (x.shape[-1], self.features), self.param_dtype)
+   bias = self.param("bias", self.bias_init, (self.features,), self.param_dtype)
+   x, kernel, bias = promote_arrays(x, kernel, bias, dtype=self.dtype)
+   return x @ kernel + bias
+```
+
+
+## Half precision types
+
+Some layers don’t work with half-precision types internally. For example: The normalization layers currently compute mean and variance in float32 even when a half precision type is specified to avoid numerical issues. We can replicate this behavior by calling result\_dtype with a dummy argument that has the minimum precision for the sub computation to work correctly.
```

Comment:
I suppose JAX also has a utility function to find the minimal dtype in the lattice? It may be useful mentioning that here. Or can we simply use `jax.results_type`?